README
This is makepatch, a program to generate a patch file from two files
or directories.

It resembles "diff -c -r -N", but:

  - it is always recursive;
  - it handles 'patchlevel.h' first;
  - it supplies 'Index:' and 'Prereq:' lines;
  - it can use 'manifest' files;
  - it generates shell commands to create and remove files and directories;
  - it can manipulate manifest files.
  - it can exclude files using wildcard and regex patters.

As distributed, it assumes that perl resides in /usr/local/bin. You
have to change the first line of makepatch.pl if your perl is in a
different location.

To install: 

  - Edit Makefile.PL (if needed)
  - Type 'perl Makefile.PL'
  - Type 'make'

    blib/script/makepatch is the makepatch program
    blib/man1/makepatch.1 is the manual page

  - Type 'make install' to install the program and the documentation
    on your system. This usually requires super user privileges.

--------------------------------------------------------------------------
Johan Vromans                                         jvromans@squirrel.nl
Squirrel Consultancy                              Haarlem, the Netherlands
http://www.squirrel.nl              http://www.squirrel.nl/people/jvromans
PGP Key 2048/4783B14D KFP=65 44 CA 66 B3 50 0B 34  CE 0E FB CA 2D 95 34 D0
---------------------- "Arms are made for hugging" -----------------------


CHANGES
Changes in 2.04
---------------

General

 * More robust determination of system locations for temp, home and
   nul.

 * Remove doc for non-existent CLO --generate.

 * More MsWin line end issues.

Changes in 2.03
---------------

General

 * More MsWin issues.

Changes in 2.02
---------------

General

 * More File::Spec issues.

 * More MsWin issues.

Changes in 2.01
---------------

General

 * Used IO::File instead of IO.

 * Use binmode for all files, to prevent unicode problems with newer
   perls. 

 * Modified a few things that required Perl 5.005, so it now runs
   under 5.004.

 * Calculate separate checksums for the patch data and the whole patch
   file. Do not complain if the checksum for the patch file is wrong,
   if the checksum for the patch data is okay.
   This allows modification of the preamble of the patch file without
   affecting the integrity checking mechanism.

 * Restructured the files in the kit, and added a test program.

 * A warning is given when installing on Solaris, since the 'patch'
   program on Solaris (at least the versions I have access to) is
   buggy to the extent that it is unusable.

 * Small fixes to avoid warnings under Perl 5.8.x and later.

Makepatch

 * Some systems need to set environment variable TMPDIR (or TEMP) to
   designate a location for temporary files.

Overall changes in makepatch version 2.0
----------------------------------------

General

 * The makepatch package now consists of two programs: 'makepatch' and
   'applypatch'. 'makepatch' generates the patch kit (using the 'diff'
   program), 'applypatch' applies it (using the 'patch' program).

Makepatch

 * To avoid most common problems with buggy versions of the 'patch'
   program, the patch data is relocated to the current directory. 
   This obsoletes the command line options -fixpath and -fixallpath.

   As a consequence, when a patch is generated from two directories,
   e.g.:

     makepatch src-1.0 src-1.1 > the-patch

   it must be applied *IN* the directory src-1.0, e.g.

     cd src-1.0; applypatch the-patch

   When fed to patch directly, no -p option should be used:

     cd src-1.0; patch < the-patch

 * (Initial) program options can be specified in environment variable
   MAKEPATCHINIT, and in preferences files /etc/makepatchrc,
   $HOME/.makepatchrc and ./.makepatchrc. The latter file name can be
   overridden with a '-rcfile' option in MAKEPATCHINIT.

 * Makepatch can take standard source archives (.tar.gz, .tar.bz2,
   .tar or .zip) as input instead of the sources directories.
   The -extract command line option can be used to change or add
   extraction rules.

 * MANIFEST files are automatically used unless suppressed with the
   new -nomanifest option.

 * Files can be excluded from the process with the new command line
   options -exclude, -exclude_regex, -exclude-vc.

Applypatch

 * The generated patch kit is the concatenation of
    - a shell preamble;
    - the actual 'patch' input;
    - a data section for 'applypatch'.

 * When applied through 'applypatch', the patch kit will be
   extensively verified for integrity. Heuristics are applied to the
   current directory to verify that it is indeed the expected source
   directory. This directory is prepared for the patching, i.e. new
   files and directories are created if required. Next, the patch data
   is passed to the 'patch' program for the actual patching. Finally
   file modes and timestamps are adjusted.

   Example:
     applypatch -d src-1.0 the-patch

 * When the patch kit is run as a shell script, it will prepare the
   current directory by creating / removing files and directories as
   required. The patch kit should then be fed to the 'patch' program.

   Example:
     cd src-1.0; sh < the-patch; patch < the-patch

 * When the patch kit is fed to the 'patch' program, the files in the
   current directory will be patched. If the directory has not been
   prepared (e.g. by running the kit as a shell script) the results
   will very much depend on your 'patch' program.

   Example:
     patch -d src-1.0 < the-patch



Makefile.PL
# Makefile.PL -- Makefile for makepatch
# Author          : Ulrich Pfeifer
# Created On      : Mon Feb 17 10:51:47 1997
# Last Modified By: Johan Vromans
# Last Modified On: Mon May 11 11:43:50 2009
# Update Count    : 106
# Status          : Released

# Verify perl version.
require 5.004;

# Verify CORE modules.
use Getopt::Long 2.00;
use IO qw(File);
use File::Basename;
use File::Spec;
use Config;

use ExtUtils::MakeMaker;

# WriteMakefile parameter hash.
my %p = ( NAME => 'makepatch', VERSION => "2.04" );

# Extra info for newer versions.
if ( $[ >= 5.005 ) {
    $p{AUTHOR} = 'Johan Vromans (jvromans@squirrel.nl)';
    $p{ABSTRACT} = 'patchkit generate and apply tool';
}

# Scripts.
my @scripts = qw (makepatch applypatch);

my $usrbin = "/usr/bin";
my $installscript = $Config{installscript};

print STDERR ("\n",
	      "WARNING: This Makefile will install user accessible scripts.\n");
print STDERR ("The location for these scripts is $installscript.\n")
  unless $installscript eq $usrbin;

print STDERR ("You may consider to pass INSTALLSCRIPT=$usrbin (or some other\n",
	      "convenient location) to \"make install\".\n\n");

if ( $^O eq "solaris" ) {
    print STDERR <<EOD;

I see you're running Solaris.
The Solaris version of the 'patch' program is extremely buggy.
If the 'make test' command fails with a message

  patch: Line must begin with '+ ', '  ', or '! '.

you must install a better version of 'patch', for example, GNU patch.

EOD
}

my $TMPDIR = $ENV{TMPDIR} || $ENV{TEMP};
$TMPDIR = File::Spec->tmpdir if !$TMPDIR && File::Spec->can("tmpdir");
$TMPDIR ||= "/usr/tmp";
unless ( -d $TMPDIR && -w $TMPDIR ) {
print STDERR <<EOD;
The default location for temporary files ($TMPDIR) does not
exist, or is not accessible.
You must use environment variable TMPDIR or TEMP to designate a writable
directory to hold temporary files.
EOD
    die ("Cannot continue\n");
}


=ignore

# Included modules.
my @mods = ();

# Required modules that are included in this kit for convenience.
# Syntax and semantics like PREREQ_PM.
my %req = ( Text::Filter => 0,
	    Text::Diff   => 0,
	    Text::Patch  => 0,
	  );

# Check modules, and ask to install them now.
$p{PREREQ_PM} = {};

foreach ( sort keys %req ) {
    my $mod = $_;
    my $rev = $req{$mod};
    my $eval = "use $mod $rev";
    eval $eval;
    if ( $@ ) {
	print STDERR ("\nI need the $mod package",
		      ($rev ? ", version $rev or higher" : ""), ".\n",
		      "Shall I install it for you? ");
	my $ans = <STDIN>;
	if ( $ans =~ /^y/i ) {
	    $mod =~ s/::/\//g;
	    push (@mods, "$mod.pm");
	}
	else {
	    # Have MakeMaker complain.
	    $p{PREREQ_PM}->{$mod} = 0;
	}
    }
}

=cut

# Append scripts and modules to WriteMakefile args.
$p{EXE_FILES} = [ map { "script/$_" } @scripts ];
foreach ( @mods ) {
    $p{PM}->{"lib/$_"} = '${INST_LIBDIR}'."/$_";
}

# Write the Makefile.
WriteMakefile (%p);
